Introduce `chain-spec-generator` and document Weight Generation process (#127)
polkadot-fellows/runtimes: README.md

2. Use it to build a chain spec for your runtime, e.g. `./target/production/chain-spec-generator --raw polkadot-local > polkadot-chain-spec.json`
3. Create `file_header.txt`

```text
// Copyright (C) Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

4. `rsync` chain spec/s and the file header to a benchmark machine

5. Build `polkadot-sdk` with `--profile production --features runtime-benchmarks` on the benchmark machine

6. Create output directories for the weights on the benchmark machine

7. Run on the benchmark machine:

```bash
for pallet in \
  frame_system \

>   # other pallets you want to benchmark

Reviewer: @liamaharon Maybe it's worth mentioning a call to list all pallets in the runtime, something like this (inspired by command-bot-scripts):
```
local pallets=($(
    ./target/production/polkadot benchmark pallet --list  |\
      --chain=/path/to/chain-spec.json |\
      tail -n+2 |\
      cut -d',' -f1 |\
      sort |\
      uniq
))

for pallet in "${pallets[@]}"; do
    echo "   [+] $pallet"
done
```

Author: Didn't know about that command, much better! Thanks 👌

  pallet_proxy; do
  echo "Running benchmark for $pallet"
  ./target/production/polkadot benchmark pallet \
    --chain=/path/to/chain-spec.json \
    --steps 50 \
    --repeat 20 \
    --pallet=$pallet \
    --extrinsic=* \
    --wasm-execution=compiled \
    --heap-pages=4096 \
    --output /path/to/runtime/weights/directory \
    --header /path/to/file_header.txt
done
```

You probably want to do this inside a `tmux` session or similar, as it will take a while.

7a. If benchmarking `pallet_alliance`

> Rename `fn add_scrupulous_items` to `fn add_unscrupulous_items` (see `https://github.com/paritytech/polkadot-sdk/pull/2173`).

Reviewer 2: Rename where? And this should also be fixed with the next sdk release or?

Author: In the weight file.

Yes it should be fixed next release. I'll just remove this note because it's unlikely anyone will want to bench this before next release.

8. `rsync` the weights back to your local machine

## FAQ

### What benchmark machine spec should I use?

> Google Cloud `n2-standard-8` or equivalent.

Reviewer 2: We should link the standard hardware in the wiki.

### Why not use `--pallet=*` when generating benchmarks?

XCM benchmarks are broken until runtimes repo gets <https://github.com/paritytech/polkadot-sdk/pull/2288>. Once this is fixed, we can use `--pallet=*` instead of a list of pallets.

### Why is this such a manual task?

It shouldn't be. Now that we have a process to follow, it should be automated by a script that takes as input:

1. List of runtimes & pallets to bench
2. SSH credentials for a benchmark machine
3. Output dir

> and writes the weights to the local output dir.

Reviewer 2: This is an issue and should not be part of the readme.

Author: Will replace with https://github.com/polkadot-fellows/runtimes/issues/128

The release process is building all runtimes and then puts them into a release in this github repository.

The format of [`CHANGELOG.md`](CHANGELOG.md) is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

> # Weight Generation

Reviewer 3: Maybe worth putting in its own file in the `docs` folder and liking it here.

To generate weights for a runtime

1. Build `chain-spec-generator` with `--profile production --features runtime-benchmarks`
2. Use it to build a chain spec for your runtime, e.g. `./target/production/chain-spec-generator --raw polkadot-local > polkadot-chain-spec.json`
3. Create `file_header.txt`

```text

> // Copyright (C) Parity Technologies (UK) Ltd.

Reviewer 3: Parity does not own the runtimes, but not sure what to put else. `Polkadot Fellowship` maybe?  
We should probably ask the lawyers who should shold the copyrights to this.

// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

4. `rsync` chain spec/s and the file header to a benchmark machine

> 5. Build `polkadot-sdk` with `--profile production --features runtime-benchmarks` on the benchmark machine

Reviewer: ```suggestion
5. Build `polkadot` binary from the latest release of `polkadot-sdk` with `--profile production --features runtime-benchmarks --bin polkadot` on the benchmark machine
```

I would specify here "latest release version" as Basti suggested [here](https://github.com/polkadot-fellows/runtimes/issues/50#issuecomment-1738925780). And also that just `--bin polkadot` is needed (not to waste resources by building the whole `polkadot-sdk` repo)

> 6. Create output directories for the weights on the benchmark machine

Reviewer 3: How are they called? Maybe the script below can also create them.

Author: added dir creation to the script and removed this step

7. Run on the benchmark machine:

```bash
#!/bin/bash

# Default value is 'polkadot', but you can override it by passing a different value as an argument
CHAIN=${1:-polkadot}

pallets=($(
  ./target/production/polkadot benchmark pallet --list \
    --chain=./$CHAIN-chain-spec.json |
    tail -n+2 |
    cut -d',' -f1 |
    sort |
    uniq
);

for pallet in "${pallets[@]}"; do
  echo "Running benchmarks for $pallet"
  ./target/production/polkadot benchmark pallet \
    --chain=./$CHAIN-chain-spec.json \
    --steps=50 \
    --repeat=20 \
    --pallet=$pallet \
    --extrinsic=* \
    --wasm-execution=compiled \
    --heap-pages=4096 \
    --output=./$CHAIN-weights \
    --header=./file_header.txt
done
```

You probably want to do this inside a `tmux` session or similar, as it will take a while.

8. `rsync` the weights back to your local machine, replacing the existing weights.

9. Manually fix XCM weights by
- Replacing `impl<T: frame_system::Config> xxx::yyy::WeightInfo<T> for WeightInfo<T> {` with `impl<T: frame_system::Config> WeightInfo<T> {`
- Marking all functions `pub(crate)`
- Removing any unused functions

10. Commit the weight changes.

> 11. If not installed, `cargo install subweight`, and check the weight changes with `subweight compare commits --path-pattern "./relay/polkadot/src/weights/*.rs" --method asymptotic --ignore-errors HEAD HEAD^1`. Ensure the changes are reasonable.

Reviewer 3: ```suggestion
11. If not installed, `cargo install subweight`, and check the weight changes with `subweight compare commits --path-pattern "./**/weights/*.rs" --method asymptotic --ignore-errors HEAD origin/main`. Ensure the changes are reasonable.
```
So that it compares all chains, in case more than just Polkadot was updated.

// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

4. `rsync` chain spec/s and the file header to a benchmark machine

5. Build `polkadot` binary from the latest release of `polkadot-sdk` with `--profile production --features runtime-benchmarks --bin polkadot` on the benchmark machine

6. Create output directories for the weights on the benchmark machine

7. Run on the benchmark machine:

```bash
#!/bin/bash

# Default value is 'polkadot', but you can override it by passing a different value as an argument
CHAIN=${1:-polkadot}

pallets=($(
  ./target/production/polkadot benchmark pallet --list \
    --chain=./$CHAIN-chain-spec.json |
    tail -n+2 |
    cut -d',' -f1 |
    sort |
    uniq

> );

Reviewer: ```suggestion
));
```

```
./bench.sh: line 13: syntax error near unexpected token `;'
./bench.sh: line 13: `);'

```

for pallet in "${pallets[@]}"; do

>   echo "Running benchmarks for $pallet"

Reviewer: ```suggestion
  output_file=./$CHAIN-weights
  # a little hack for pallet_xcm_benchmarks - we want to force custom implementation for XcmWeightInfo
  if [[ "$pallet" == "pallet_xcm_benchmarks::generic" ]] || [[ "$pallet" == "pallet_xcm_benchmarks::fungible" ]]; then
    output_file="$output_file/xcm/${pallet//::/_}.rs"
  fi
  echo "Running benchmarks for $pallet to $output_file"
```

  ./target/production/polkadot benchmark pallet \
    --chain=./$CHAIN-chain-spec.json \
    --steps=50 \
    --repeat=20 \
    --pallet=$pallet \
    --extrinsic=* \
    --wasm-execution=compiled \
    --heap-pages=4096 \

>     --output=./$CHAIN-weights \

Reviewer: ```suggestion
    --output=$output_file \
```

usually, there is hack for xcm benchmarks, which are placed under `*/weights/xcm`:
```
# a little hack for pallet_xcm_benchmarks - we want to force custom implementation for XcmWeightInfo
    if [[ "$pallet" == "pallet_xcm_benchmarks::generic" ]] || [[ "$pallet" == "pallet_xcm_benchmarks::fungible" ]]; then
      output_file="xcm/${pallet//::/_}.rs"
      extra_args="--template=$output_path/templates/xcm-bench-template.hbs"
    fi
```
